File: src/agents/CollectorAgent.ts
import { ContentItem, AgentResponse } from '../types';

export class CollectorAgent {
  private static instance: CollectorAgent;

  static getInstance(): CollectorAgent {
    if (!CollectorAgent.instance) {
      CollectorAgent.instance = new CollectorAgent();
    }
    return CollectorAgent.instance;
  }

  async extractFromPDF(file: File): Promise<AgentResponse> {
    const startTime = Date.now();
    
    try {
      // Mock PDF text extraction
      const mockText = `
        Project Planning Document
        
        Objective: Complete the quarterly marketing campaign
        
        Key Components:
        1. Market research and competitor analysis
        2. Creative asset development
        3. Campaign timeline and budget allocation
        4. Performance metrics and KPIs
        
        Resources needed:
        - Design team collaboration
        - Budget approval from finance
        - Content creation timeline
        - Distribution channel strategy
      `;

      const contentItem: ContentItem = {
        id: `pdf-${Date.now()}`,
        type: 'pdf',
        title: file.name,
        content: file.name,
        extractedText: mockText,
        metadata: {
          fileSize: file.size,
          lastModified: file.lastModified,
          mimeType: file.type
        },
        timestamp: new Date(),
        status: 'completed',
        categories: []
      };

      return {
        success: true,
        data: contentItem,
        processingTime: Date.now() - startTime
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        processingTime: Date.now() - startTime
      };
    }
  }

  async extractFromImage(file: File): Promise<AgentResponse> {
    const startTime = Date.now();
    
    try {
      // Mock OCR text extraction
      const mockText = `
        Meeting Notes - Project Kickoff
        Date: December 15, 2024
        
        Attendees: Sarah, Mike, Alex, Jennifer
        
        Agenda Items:
        • Project scope and deliverables
        • Timeline and milestones
        • Resource allocation
        • Risk assessment
        
        Action Items:
        1. Sarah - Finalize project requirements by Dec 20
        2. Mike - Set up development environment
        3. Alex - Create initial wireframes
        4. Jennifer - Schedule stakeholder reviews
        
        Next Meeting: December 22, 2024
      `;

      const contentItem: ContentItem = {
        id: `img-${Date.now()}`,
        type: 'image',
        title: file.name,
        content: file.name,
        extractedText: mockText,
        metadata: {
          fileSize: file.size,
          dimensions: '1920x1080', // Mock dimensions
          format: file.type
        },
        timestamp: new Date(),
        status: 'completed',
        categories: []
      };

      return {
        success: true,
        data: contentItem,
        processingTime: Date.now() - startTime
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        processingTime: Date.now() - startTime
      };
    }
  }

  async extractFromYouTube(url: string): Promise<AgentResponse> {
    const startTime = Date.now();
    
    try {
      // Mock YouTube transcript extraction
      const mockTranscript = `
        Welcome to this tutorial on project management best practices.
        
        Today we'll cover:
        - Setting clear objectives and scope
        - Building effective team communication
        - Managing timelines and deadlines
        - Risk mitigation strategies
        - Quality assurance processes
        
        The first step in any successful project is defining clear, measurable objectives.
        Without proper goal setting, teams often lose focus and deliverables become unclear.
        
        Communication is the backbone of project success. Regular check-ins, status updates,
        and transparent reporting help keep everyone aligned and accountable.
        
        Timeline management requires balancing optimism with realism. Build in buffer time
        for unexpected challenges while maintaining momentum toward key milestones.
      `;

      const contentItem: ContentItem = {
        id: `yt-${Date.now()}`,
        type: 'youtube',
        title: 'Project Management Best Practices',
        content: url,
        extractedText: mockTranscript,
        metadata: {
          url,
          duration: '15:32',
          publishedDate: '2024-12-01',
          channelName: 'PM Academy'
        },
        timestamp: new Date(),
        status: 'completed',
        categories: []
      };

      return {
        success: true,
        data: contentItem,
        processingTime: Date.now() - startTime
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        processingTime: Date.now() - startTime
      };
    }
  }

  async extractFromText(text: string, title: string = 'Text Note'): Promise<AgentResponse> {
    const startTime = Date.now();
    
    try {
      const contentItem: ContentItem = {
        id: `text-${Date.now()}`,
        type: 'text',
        title,
        content: text,
        extractedText: text,
        metadata: {
          wordCount: text.split(' ').length,
          characterCount: text.length
        },
        timestamp: new Date(),
        status: 'completed',
        categories: []
      };

      return {
        success: true,
        data: contentItem,
        processingTime: Date.now() - startTime
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        processingTime: Date.now() - startTime
      };
    }
  }
}

File: backend/agents/collector_agent.py
import asyncio
from typing import Dict, Any, List
import uuid
from datetime import datetime
import os
import PyPDF2
import pytesseract
from PIL import Image
import requests
from youtube_transcript_api import YouTubeTranscriptApi
import re

# Try to import AI libraries for enhanced processing
try:
    import google.genai as genai
    from google.genai import types
    HAS_GEMINI = True
except ImportError:
    HAS_GEMINI = False
    print("⚠️ google-genai not available, using fallback implementations")

class CollectorAgent:
    """Agent responsible for collecting and extracting content from various sources using Gemini AI for enhancement"""
    
    def __init__(self):
        self.name = "CollectorAgent"
        self.description = "Extracts and processes content from PDFs, images, YouTube videos, and text with Gemini AI enhancement"
        
        # Initialize Gemini client for content enhancement
        self.gemini_client = None
        if HAS_GEMINI and os.getenv('GOOGLE_API_KEY'):
            try:
                self.gemini_client = genai.Client(api_key=os.getenv('GOOGLE_API_KEY'))
                print("✅ Gemini AI client initialized for content enhancement")
            except Exception as e:
                print(f"⚠️ Gemini initialization failed: {e}")
        
        print(f"✅ Initialized {self.name}")
    
    async def extract_from_pdf(self, file_path: str) -> Dict[str, Any]:
        """Extract text content from PDF files using PyPDF2 with Gemini enhancement"""
        start_time = datetime.now()
        
        try:
            extracted_text = ""
            
            # Real PDF extraction using PyPDF2
            try:
                with open(file_path, 'rb') as file:
                    pdf_reader = PyPDF2.PdfReader(file)
                    for page_num in range(len(pdf_reader.pages)):
                        page = pdf_reader.pages[page_num]
                        extracted_text += page.extract_text() + "\n"
            except Exception as pdf_error:
                print(f"⚠️ PDF extraction failed, using fallback: {str(pdf_error)}")
                # Fallback to mock data if PDF extraction fails
                extracted_text = """
                Project Planning Document
                
                Objective: Complete the quarterly marketing campaign
                
                Key Components:
                1. Market research and competitor analysis
                2. Creative asset development
                3. Campaign timeline and budget allocation
                4. Performance metrics and KPIs
                
                Resources needed:
                - Design team collaboration
                - Budget approval from finance
                - Content creation timeline
                - Distribution channel strategy
                """
            
            # If no text extracted, use fallback
            if not extracted_text.strip():
                extracted_text = """
                Project Planning Document
                
                Objective: Complete the quarterly marketing campaign
                
                Key Components:
                1. Market research and competitor analysis
                2. Creative asset development
                3. Campaign timeline and budget allocation
                4. Performance metrics and KPIs
                
                Resources needed:
                - Design team collaboration
                - Budget approval from finance
                - Content creation timeline
                - Distribution channel strategy
                """
            
            # Enhance with Gemini AI if available
            enhanced_text = await self._enhance_extracted_text(extracted_text, "PDF document")
            
            content_item = {
                "id": f"pdf-{uuid.uuid4()}",
                "type": "pdf",
                "title": os.path.basename(file_path),
                "content": file_path,
                "extracted_text": enhanced_text or extracted_text.strip(),
                "metadata": {
                    "file_path": file_path,
                    "file_size": os.path.getsize(file_path) if os.path.exists(file_path) else 0,
                    "pages": len(pdf_reader.pages) if 'pdf_reader' in locals() else 0,
                    "enhanced_by_ai": bool(enhanced_text)
                },
                "timestamp": datetime.now().isoformat(),
                "status": "completed",
                "categories": []
            }
            
            print(f"📄 Extracted {len(content_item['extracted_text'])} characters from PDF: {os.path.basename(file_path)}")
            
            return {
                "success": True,
                "data": content_item,
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
            
        except Exception as e:
            print(f"❌ Error extracting PDF: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
    
    async def extract_from_image(self, file_path: str) -> Dict[str, Any]:
        """Extract text from images using OCR (Tesseract) with Gemini enhancement"""
        start_time = datetime.now()
        
        try:
            extracted_text = ""
            
            # Real OCR extraction using Tesseract
            try:
                image = Image.open(file_path)
                extracted_text = pytesseract.image_to_string(image)
            except Exception as ocr_error:
                print(f"⚠️ OCR failed, using fallback: {str(ocr_error)}")
                # Fallback to mock data if OCR fails
                extracted_text = """
                Meeting Notes - Project Kickoff
                Date: December 15, 2024
                
                Attendees: Sarah, Mike, Alex, Jennifer
                
                Agenda Items:
                • Project scope and deliverables
                • Timeline and milestones
                • Resource allocation
                • Risk assessment
                
                Action Items:
                1. Sarah - Finalize project requirements by Dec 20
                2. Mike - Set up development environment
                3. Alex - Create initial wireframes
                4. Jennifer - Schedule stakeholder reviews
                
                Next Meeting: December 22, 2024
                """
            
            # Enhance with Gemini AI if available
            enhanced_text = await self._enhance_extracted_text(extracted_text, "image/screenshot")
            
            content_item = {
                "id": f"img-{uuid.uuid4()}",
                "type": "image",
                "title": os.path.basename(file_path),
                "content": file_path,
                "extracted_text": enhanced_text or extracted_text.strip(),
                "metadata": {
                    "file_path": file_path,
                    "file_size": os.path.getsize(file_path) if os.path.exists(file_path) else 0,
                    "format": "image",
                    "enhanced_by_ai": bool(enhanced_text)
                },
                "timestamp": datetime.now().isoformat(),
                "status": "completed",
                "categories": []
            }
            
            print(f"🖼️ Extracted {len(content_item['extracted_text'])} characters from image: {os.path.basename(file_path)}")
            
            return {
                "success": True,
                "data": content_item,
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
            
        except Exception as e:
            print(f"❌ Error extracting from image: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
    
    async def extract_from_youtube(self, url: str) -> Dict[str, Any]:
        """Extract transcript from YouTube videos with Gemini enhancement"""
        start_time = datetime.now()
        
        try:
            extracted_text = ""
            title = "YouTube Video"
            
            # Extract video ID from URL
            video_id = self._extract_video_id(url)
            
            if video_id:
                try:
                    # Real YouTube transcript extraction
                    transcript_list = YouTubeTranscriptApi.get_transcript(video_id)
                    extracted_text = " ".join([item['text'] for item in transcript_list])
                    
                    # Try to get video title (simplified approach)
                    title = f"YouTube Video ({video_id})"
                    
                except Exception as yt_error:
                    print(f"⚠️ YouTube transcript extraction failed, using fallback: {str(yt_error)}")
                    # Fallback to mock data
                    extracted_text = """
                    Welcome to this tutorial on project management best practices.
                    
                    Today we'll cover:
                    - Setting clear objectives and scope
                    - Building effective team communication
                    - Managing timelines and deadlines
                    - Risk mitigation strategies
                    - Quality assurance processes
                    
                    The first step in any successful project is defining clear, measurable objectives.
                    Without proper goal setting, teams often lose focus and deliverables become unclear.
                    
                    Communication is the backbone of project success. Regular check-ins, status updates,
                    and transparent reporting help keep everyone aligned and accountable.
                    
                    Timeline management requires balancing optimism with realism. Build in buffer time
                    for unexpected challenges while maintaining momentum toward key milestones.
                    """
                    title = "Project Management Best Practices"
            else:
                # Invalid URL, use fallback
                extracted_text = "Invalid YouTube URL provided."
                title = "Invalid YouTube Video"
            
            # Enhance with Gemini AI if available
            enhanced_text = await self._enhance_extracted_text(extracted_text, "YouTube transcript")
            
            content_item = {
                "id": f"yt-{uuid.uuid4()}",
                "type": "youtube",
                "title": title,
                "content": url,
                "extracted_text": enhanced_text or extracted_text.strip(),
                "metadata": {
                    "url": url,
                    "video_id": video_id,
                    "duration": "Unknown",
                    "published_date": "Unknown",
                    "channel_name": "Unknown",
                    "enhanced_by_ai": bool(enhanced_text)
                },
                "timestamp": datetime.now().isoformat(),
                "status": "completed",
                "categories": []
            }
            
            print(f"🎥 Extracted {len(content_item['extracted_text'])} characters from YouTube: {title}")
            
            return {
                "success": True,
                "data": content_item,
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
            
        except Exception as e:
            print(f"❌ Error extracting from YouTube: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
    
    async def extract_from_text(self, text: str, title: str = "Text Note") -> Dict[str, Any]:
        """Process plain text input with Gemini enhancement"""
        start_time = datetime.now()
        
        try:
            # Enhance with Gemini AI if available
            enhanced_text = await self._enhance_extracted_text(text, "text note")
            
            content_item = {
                "id": f"text-{uuid.uuid4()}",
                "type": "text",
                "title": title,
                "content": text,
                "extracted_text": enhanced_text or text,
                "metadata": {
                    "word_count": len(text.split()),
                    "character_count": len(text),
                    "enhanced_by_ai": bool(enhanced_text)
                },
                "timestamp": datetime.now().isoformat(),
                "status": "completed",
                "categories": []
            }
            
            print(f"📝 Processed text: {title} ({len(content_item['extracted_text'])} characters)")
            
            return {
                "success": True,
                "data": content_item,
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
            
        except Exception as e:
            print(f"❌ Error processing text: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "processing_time": (datetime.now() - start_time).total_seconds()
            }
    
    async def _enhance_extracted_text(self, text: str, content_type: str) -> str:
        """Use Gemini AI to enhance and clean up extracted text"""
        try:
            if not self.gemini_client or len(text.strip()) < 50:
                return None
            
            prompt = f"""
            Clean up and enhance the following extracted text from a {content_type}.
            Fix any OCR errors, improve formatting, and make it more readable while preserving all original meaning and information.
            Do not add new information, only clean up and structure what's already there.
            
            Original text:
            {text[:2000]}...
            
            Return the cleaned and enhanced version:
            """
            
            response = await asyncio.to_thread(
                self.gemini_client.models.generate_content,
                model='gemini-2.0-flash-exp',
                contents=prompt,
                config=types.GenerateContentConfig(
                    temperature=0.1,
                    max_output_tokens=2000
                )
            )
            
            enhanced_text = response.text.strip()
            print(f"✨ Enhanced {content_type} text with Gemini AI")
            return enhanced_text
            
        except Exception as e:
            print(f"⚠️ Text enhancement failed: {e}")
            return None
    
    def _extract_video_id(self, url: str) -> str:
        """Extract video ID from YouTube URL"""
        patterns = [
            r'(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\n?#]+)',
            r'youtube\.com\/watch\?.*v=([^&\n?#]+)'
        ]
        
        for pattern in patterns:
            match = re.search(pattern, url)
            if match:
                return match.group(1)
        
        return None